src/neighborsearch/neighborhood.jl

```diff
@@ -63,8 +76,69 @@ function search(xₒ::AbstractVector, method::NeighborhoodSearch{O,N,T};
         push!(neighbors, ind)
       end
     end
+    method.restrictions && (neighbors = filterneighs(neighbors, method, xₒ))
     neighbors
   else
+    method.restrictions && (inds = filterneighs(inds, method, xₒ))
     inds
   end
 end
+
+function filterneighs(inds::AbstractVector, method::NeighborhoodSearch{O,N,T},
+  xₒ::AbstractVector) where {O,N<:BallNeighborhood,T}
+
+  object   = method.object
+  neigh    = method.neigh
+  rotmat   = neigh.metric isa Mahalanobis ? neigh.metric.qmat : I
+  ometric  = method.ordermetric
+  maxk     = method.maxneighs
+  maxoct   = method.maxperoct
+  maxkey   = method.maxperkey
+  usek     = 0 < maxk < length(inds)
+  useoct   = maxoct > 0
+  usekey   = maxkey[2] > 0
+
+  # get distances and give priority to closest neighbors according to ordermetric
+  dists   = colwise(ometric, xₒ, coordinates(object,inds))
+  sortids = sortperm(dists)
+  inds    = inds[sortids]
+
+  # key and octant constraints
+  keys   = usekey ? unique(view(values(object), inds, maxkey[1])) : nothing
+  dkeys  = usekey ? Dict(zip(keys, 1:length(keys)))               : nothing
+  ctkeys = usekey ? zeros(Int, length(keys))                      : nothing
+  octs   = useoct ? zeros(Int, 8)                                 : nothing
+
+  neighbors = Vector{Int}()
+  ctneighs  = 0
+
+  @inbounds for ind in inds
+    key = usekey ? values(object)[ind,maxkey[1]]                    : nothing
+    oct = useoct ? getoct(rotmat * (coordinates(object,ind) .- xₒ)) : nothing
+
+    usekey && ctkeys[dkeys[key]] >= maxkey[2] && continue
+    useoct && octs[oct] >= maxoct             && continue
```

Comment:
So this is where the filtering is happening, right? It is very hard to understand what exactly is going on, and so I suggest to add some comments to explain what your intentions are with the code written. I like the formatting btw, very nice to see a PR with with everything aligned properly 💯